### DISCUSSION TITLE
[@types/react] 18 ' Fragment children not accept multiple children

### DISCUSSION THREAD

**Discussion by cqh963852** [DefinitelyTyped/DefinitelyTyped#59816] [Issues with a @types package] [answered]
![image](https://user-images.githubusercontent.com/17702287/162680624-71680852-d2af-47d6-8652-786d04d842bb.png)

In react 17, Fragment can accept multiple children, and works fine.

but in react 18 ' Fragment children not accept multiple children.

It is an type bug? 
Or is there a new way to practice?

[a comment by an automated account is not shown]

**Reply by cqh963852** [reply]
https://reactjs.org/docs/fragments.html

![image](https://user-images.githubusercontent.com/17702287/162682396-9084089c-6c92-4ab2-bd57-855abb00f4c6.png)

While use Fragment with out a map. The error gone. 

Can any one help?

**Reply by cqh963852** [reply]
![image](https://user-images.githubusercontent.com/17702287/162682746-5a09825c-6c79-4ab8-b8a7-53cba24ffaef.png)
Sorry ,I forget return in map function.

**Comment by ferdaber** [comment]
I wonder if it’s because the Fragment component was dependent on implicit children? The workaround for now would be to use `<></>` shorthand

**Reply by eps1lon** [reply]
You can no longer assign `void[]` to `ReactNode` which is what was happening in the original repro: 
```tsx
import * as React from "react";

function Menu() {
  return null;
}

function AppA({ menus }: { menus?: void[] }) {
  return (
    // This JSX tag's 'children' prop expects a single child of type 'ReactNode', but multiple children were provided.(2746)
    <React.Fragment>
      <div />
      {menus}
    </React.Fragment>
  );
}

function AppB({ menus }: { menus?: void[] }) {
    // Type 'void[] | undefined' is not assignable to type 'ReactNode'.
  return <React.Fragment>{menus}</React.Fragment>;
}
```

[Playground Link](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcARFDvmQNwBQdBArgHb7AQtwCyWLTACgCUcAN504cSjCZQu-ADYL6AXwbM2MDlwCCYMDoGi4IPk3QqAXGJNnUAfmsA3CMAAmAbQC6cFSPGS0rJcAhKScAD0EXAAKgAWwOgAUgDKABpwMMgA5gDk6Lm4CQpulCy5cGDEYHBYAB5gWPjoyHCowCzZClhwRcAlcBAEmQCejXC52HgwAHIQbli5ADRwAEZM8CBMClpg3b3FpXxwAO5YlJXETu5YbgB0AgBMAOwALABsQmGSADxT+HcAGJQHKmFgwAB833CPzcwCckSh4XCojB5jUyJ+EX+MCBIOyYMhYSEqnUrHYnDgejAACEjLZ+BZrMY0Q5nK5PD4-GJoVFYmMerkXO5vHAAD5wVgLAgdW4VRJwFgQeBodrZFjIVb7GAQUbjSZUWbzRZ3MJBORwP6GvGgviQ1F2FRYnE2gl2iGkoA)

We could add `void` to the possible types of `ReactNode`. The question is if there are legitimate use-cases for `void` considering the repro was a mistake.

**Reply by ferdaber** [reply]
I feel like allowing `void` is probably fine? But I also recognize my own bias for this because I will often do early `return` statements without explicitly writing `return undefined` 🤷🏻‍♀️

**Reply by eps1lon** [reply]
Yeah if `return` means `void` then we should probably add it. I was hoping that `void` implies no `return` at all.

**Reply by eps1lon** [reply]
Continued in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/59833. TL;DR: I think the new behavior is fine since early return is still support. Only if you have no return at all or only empty returns do we no longer compile. But I think this is actually super nice.
